crates/core/src/routing_info/mod.rs

```diff
@@ -0,0 +1,219 @@
+        self.partition_to_node_mappings
+            .load()
+            .inner
+            .get(&partition_id)
+            .copied()
+    }
+
+    /// Request a refresh without waiting for it to complete. This is useful when the caller
+    /// discovers via some other mechanism that the local view might be outdated - for example, when
+    /// a request to a node previously returned by `get_node_by_partition` fails with a response
+    /// that indicates that this routing information is no longer valid.
+    pub async fn request_refresh(&self) {
+        self.sender
+            .send(Command::SyncRoutingInformation(TargetVersion::Latest))
+            .await
+            .expect("Failed to send refresh request");
+    }
+}
+
+struct PartitionToNodesRoutingTable {
+    /// A mapping of partition IDs to node IDs that are believed to be authoritative for that
+    /// serving requests for that partition.
+    inner: HashMap<PartitionId, NodeId, Xxh3Builder>,
+}
+
+/// Task to refresh the routing information, periodically or on-demand. A single
+/// instance of this component per node ensures that we don't get a stampeding
+/// herd of queries to the source of truth.
+//
+// Implementation note: the sole authority for node-level routing is currently the metadata store,
+// and in particular, the cluster scheduling plan. This will change in the future so avoid leaking
+// implementation details or assumptions about the source of truth.
+pub struct PartitionRoutingRefresher {
+    sender: CommandSender,
+    receiver: CommandReceiver,
+    metadata_store_client: MetadataStoreClient,
+    inflight_refresh_task: Option<TaskHandle<anyhow::Result<()>>>,
+    inner: Arc<ArcSwap<PartitionToNodesRoutingTable>>,
+}
+
+impl PartitionRoutingRefresher {
+    pub fn new(metadata_store_client: MetadataStoreClient) -> Self {
+        let (sender, receiver) = mpsc::channel(1);
+        Self {
+            metadata_store_client,
+            receiver,
+            sender,
+            inflight_refresh_task: None,
+            inner: Arc::new(ArcSwap::new(Arc::new(PartitionToNodesRoutingTable {
+                inner: HashMap::default(),
+            }))),
+        }
+    }
+
+    /// Get a handle to the partition-to-node routing table.
+    pub fn partition_routing(&self) -> PartitionRouting {
+        PartitionRouting {
+            sender: self.sender.clone(),
+            partition_to_node_mappings: self.inner.clone(),
+        }
+    }
+
+    async fn run(mut self) -> anyhow::Result<()> {
+        debug!("Metadata manager started");
+
+        let update_interval = Configuration::pinned()
+            .common
+            .metadata_update_interval
+            .into();
+        let mut update_interval = tokio::time::interval(update_interval);
+        update_interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        let mut cancel = pin!(cancellation_watcher());
+
+        loop {
+            tokio::select! {
```

Comment:
Is this a reasonable way to implement this? I based it off the `MetadataManager`.

Reply:
Yes, looks good to me :-)